### PR TITLE
Fix typos in RetrieveLicenseTexts.kt and the `STRING_TOO_LONG` error for CDDL license text

### PR DESCRIPTION
**Pull request by prayutsu**
## Explanation

The PR fixes the STRING_TOO_LONG error that occurs when the license text is too long to be displayed and some typos in RetrieveLicenseTexts.kt script.

Tests passing locally - 
Screenshot from 2021-08-24 13-26-24
Screenshot from 2021-08-24 13-34-00

## Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt [hunk 1]
```diff
@@ -179,44 +178,7 @@ class MavenDependenciesListCheck(
...
-        """.trimIndent()
-      )
-      License.VerifiedLinkCase.DIRECT_LINK_ONLY -> println(
-        """
-          direct_link_only: "${license.directLinkOnly.url}"
-        """.trimIndent()
-      )
+      println(dep)
     }
-    println("}")
```
Comment (prayutsu) [inline scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: These changes were done to print the artifacts with 2 spaces indentation but it didn't work. Also, I noticed in [#3684(checks)](https://github.com/oppia/oppia-android/pull/3684/checks?check_run_id=3408712501) that the error output was badly formatted.
Screenshot from 2021-08-24 11-09-57
So I fixed this and also modified the corresponding test case to ensure this isn't broken in the future.

**Comment by prayutsu** [comment]
@rt4914 @BenHenning PTAL.

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt (hunk 1 above)
Comment (rt4914) [reply scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: @prayutsu Is this PR link mentioned above is correct ?

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt [hunk 2]
```diff
@@ -20,6 +20,7 @@ import javax.xml.transform.stream.StreamResult
 
 private const val MAVEN_DEPENDENCY_LIST_NOT_UP_TO_DATE =
   "maven_dependencies.textproto is not up-to-date"
+private const val MAX_CHARS_LIMIT = 16383
```
Comment (rt4914) [inline scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Can you mention the reference from where you got this limit?

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt [hunk 3]
```diff
@@ -30,12 +30,15 @@ class RetrieveLicenseTextsTest {
   private val MAVEN_DEPENDENCY_LIST_NOT_UP_TO_DATE_FAILURE =
     "maven_dependencies.textproto is not up-to-date"
   private val SCRIPT_PASSED_INDICATOR = "Script execution completed successfully."
+  private val LONG_LICENSE_TEXT_LENGTH = 16384
```
Comment (rt4914) [inline scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: I think we should fetch this value from `RetrieveLicenseTexts.kt` file.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt [hunk 4]
```diff
@@ -539,6 +549,14 @@ class RetrieveLicenseTextsTest {
     return itemList
   }
 
+  private fun retrieveLongLicenseText(): String {
+    val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9') + ' '
```
Comment (rt4914) [inline scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: I am unable to understand this. Are we not considering any special characters? If no, then why?

**Review by rt4914 — CHANGES_REQUESTED** [review]
@prayutsu Mostly clarifications.

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt (hunk 1 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: The PR link was correct but I have updated it to point to the specific check failure.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 2 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: [This](https://stackoverflow.com/a/51733275/12215015) answer says that the max limit is 32,767 bytes and a char takes 2 bytes. So I halved the mentioned size and set it as the limit.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt (hunk 3 above)
Comment (prayutsu) [reply scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt (hunk 4 above)
Comment (prayutsu) [reply scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: Changes this to include all chars from ASCII value 32 to 122.

**Comment by prayutsu** [comment]
@rt4914 PTAL followed up on all the comments.

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt [hunk 5]
```diff
@@ -179,44 +178,7 @@ class MavenDependenciesListCheck(
...
-        """
-          extracted_copy_link: "${license.extractedCopyLink.url}"
-        """.trimIndent()
-      )
-      License.VerifiedLinkCase.DIRECT_LINK_ONLY -> println(
-        """
-          direct_link_only: "${license.directLinkOnly.url}"
-        """.trimIndent()
-      )
+      println(dep)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: Can you print the TextFormat version of the dep, instead? I don't think there's any guarantee that the Message.toString() method prints a parsable text proto (though it seems to be close).

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt [hunk 6]
```diff
@@ -147,16 +150,22 @@ class RetrieveLicenseTexts(
...
+        if (licenseText.length > MAX_CHARS_LIMIT) {
+          licenseText = licenseLink
+        }
       }
       License.VerifiedLinkCase.EXTRACTED_COPY_LINK -> {
         licenseText = fetchLicenseText(license.extractedCopyLink.url)
         licenseLink = license.extractedCopyLink.url
+        if (licenseText.length > MAX_CHARS_LIMIT) {
+          licenseText = licenseLink
+        }
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: I don't think this is what we want to do, actually. Can we instead fragment the text into multiple strings & display them in segments in the UI? We intentionally reserve the link for cases when we can't show the text, but in this case we still can with additional work. I think that's important here.

If there is some technical reason we can't do this, though, then the link might be a reasonable back-up.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @prayutsu. Took an initial pass--I think more work is needed on the large text license handling. PTAL.

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt (hunk 5 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 6 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Filed issue #3738 and added TODO.

**Comment by prayutsu** [comment]
@rt4914 @BenHenning PTAL.

File: scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt (hunk 1 above)
Comment (rt4914) [reply scripts/src/java/org/oppia/android/scripts/license/MavenDependenciesListCheck.kt]: Okay thanks.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 2 above)
Comment (rt4914) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Please mention this link in comments above the value
`Reference: https://stackoverflow.com/a/51733275`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@prayutsu One pending comment: https://github.com/oppia/oppia-android/pull/3720#discussion_r696343533

Also, please file a issue mentioning that what this current solution is doing and what are the two possible solutions which can be used in future to update this.

Also, mention that issue somewhere correctly in code where we need to update that implementation.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 2 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Done.

**Comment by prayutsu** [comment]
> @prayutsu One pending comment: [#3720 (comment)](https://github.com/oppia/oppia-android/pull/3720#discussion_r696343533)
> 
> Also, please file a issue mentioning that what this current solution is doing and what are the two possible solutions which can be used in future to update this.
> 
> Also, mention that issue somewhere correctly in code where we need to update that implementation.

Filed issue #3738

**Comment by prayutsu** [comment]
@rt4914 PTAL.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt [hunk 7]
```diff
@@ -21,6 +21,10 @@ import javax.xml.transform.stream.StreamResult
 private const val MAVEN_DEPENDENCY_LIST_NOT_UP_TO_DATE =
   "maven_dependencies.textproto is not up-to-date"
 
+/** Maximum number of chars that can be displayed in a textview. */
+// Reference: https://stackoverflow.com/a/51733275
+const val MAX_CHARS_LIMIT = 16383
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Why this number versus the actual upper limit of 32KB?

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt [hunk 8]
```diff
@@ -21,6 +21,10 @@ import javax.xml.transform.stream.StreamResult
 private const val MAVEN_DEPENDENCY_LIST_NOT_UP_TO_DATE =
   "maven_dependencies.textproto is not up-to-date"
 
+/** Maximum number of chars that can be displayed in a textview. */
+// Reference: https://stackoverflow.com/a/51733275
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Perhaps https://issuetracker.google.com/issues/110853846#comment5 would be a more authoritative reference? I think the SO link is referencing that one.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt [hunk 9]
```diff
@@ -147,16 +151,26 @@ class RetrieveLicenseTexts(
   }
 
   private fun retrieveCopyrightLicense(license: License): CopyrightLicense {
-    val licenseText: String
+    var licenseText: String
     val licenseLink: String
     when (license.verifiedLinkCase) {
       License.VerifiedLinkCase.SCRAPABLE_LINK -> {
         licenseText = fetchLicenseText(license.scrapableLink.url)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: I suggest introducing a new helper (e.g. fetchViewableLicenseText) that calls fetchLicenseText directly & does the size check there, then returns the best value.

This avoids mutable state & duplicate code between the two cases. It also lets you consolidate the TODO to one location.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt [hunk 10]
```diff
@@ -539,6 +549,14 @@ class RetrieveLicenseTextsTest {
     return itemList
   }
 
+  private fun retrieveLongLicenseText(): String {
+    val charPool: List<Char> = (' '..'z').toList() // Inlcude all chars from ASCII value 32 to 122.
+    return (1..LONG_LICENSE_TEXT_LENGTH)
+      .map { kotlin.random.Random.nextInt(0, charPool.size) }
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: Do we really need to fully qualify the class or can we import it, instead?

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 7 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Also for name, this should be specific to the content: ``MAX_LICENSE_LENGTH``.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt [hunk 11]
```diff
@@ -30,12 +30,15 @@ class RetrieveLicenseTextsTest {
...
   private val SCRIPT_PASSED_INDICATOR = "Script execution completed successfully."
+  private val LONG_LICENSE_TEXT_LENGTH = MAX_CHARS_LIMIT + 1
 
   private val SCRAPABLE_LINK = "https://www.apache.org/licenses/LICENSE-2.0.txt"
   private val DIRECT_LINK_ONLY = "https://developer.android.com/studio/terms.html"
   private val EXTRACTED_COPY_ORIGINAL_LINK = "https://www.opensource.org/licenses/bsd-license"
   private val EXTRACTED_COPY_LINK = "https://raw.githubusercontent.com/oppia/oppia-android-" +
     "licenses/develop/simplified-bsd-license.txt"
+  private val LONG_LICENSE_TEXT_LINK = "https://raw.githubusercontent.com/javaee/javax." +
+    "annotation/83417807ad402ee1022c0307208d4510c80c68b6/LICENSE"
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: If this doesn't need to be a real link, prefer to use a fake one instead.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @prayutsu. Just had a couple of comments, otherwise the PR LGTM.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt (hunk 10 above)
Comment (prayutsu) [reply scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt (hunk 11 above)
Comment (prayutsu) [reply scripts/src/javatests/org/oppia/android/scripts/maven/RetrieveLicenseTextsTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 9 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 8 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt (hunk 7 above)
Comment (prayutsu) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: I halved the 32KB size as a single char takes 2 bytes. Is it correct or is there any other thing that needs to be taken into account?

Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/maven/RetrieveLicenseTexts.kt]: Ah, good point--that sounds correct. Thanks!

**Review by BenHenning — APPROVED** [review]
Thanks @prayutsu. This LGTM!

**Pull request merged**